design/metadata-handling.md

```diff
@@ -0,0 +1,286 @@
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  values:
+    node-0:
+      abc: def
+    node-1:
+      abc: ghi
+  templates:
+    local-hostname: worker-np1-{{ getIndex }}
+```
+
+This object will be reconciled by the Metal3machine controller. When reconciled,
+the controller will add an OwnerReference to the Metal3Cluster that has nodes
+linking to this object. The spec contains two fields:
+
+- `values` contains a map of id and map of key and values. The id is merely an
+  identifier of the map of key and values.
+- `templates` contains a map of key and values that will be reused for all.
+
+In both cases, the values will be [go templates](
+https://golang.org/pkg/text/template/). A function `getIndex` will return
+the lower int that is not yet in use.
+
+A second object
+would be created, a Metal3MetadataStatus type, linked to a Metal3Cluster.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3MetadataStatus
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Metadata
+    name: nodepool-1
+spec:
+  values:
+    "node-0": default/machine-1
+  indexes:
+    "0": default/machine-1
+  machines:
+    "default/machine-1":
+      index: "0"
+      value: "node-0"
+```
+
+This object would store the list of values used for each Metal3Metadata object.
+The rationale behind not including it as a status of the previous object is to
+be able to go through the move phase without information loss. This object would
+only be used by the controller and should not be edited by the user.
+
+The name of this status object will be identical to the Metal3Metadata object.
+The `spec` field will contain fields that keeps track of which Metal3Machine is
+using each of the values or indexes.
+
+The Metal3Machine will be modified as follow:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  metaData:
+    ghi: jkl
+  metaDataTemplate:
+    name: nodepool-1
+    namespace: default
+```
+
+When CAPM3 controller will fill select a BaremetalHost and set the different
+fields there, it will generate the metaData content based on the static metaData
+and then render the metadata from the metaDataTemplate. It will
+get the referenced Metal3Metadata and render metadata from the templates,
+selecting the lowest available index, and from the values, selecting the first
+available id. It will save the index and id in the Metal3MetadataStatus. If the
+Metal3Machine already has an index and an id in the Metal3MetadataStatus, it
+will re-use those.
+
+When the Metal3Machine gets deleted, the CAPM3 controller will remove its
+references from the Metal3MetadataStatus, making it available for other
+Metal3Machines.
```

Comment:
We would only have to do this if we try to use a resource as a database. If we take some templates as input and render them into a resource that is associated one-to-one with the host, we can set the owner reference to the Metal3Machine so that when the host is deprovisioned the metadata is removed.

Reply:
that's a very good point, I will update the proposal